wagtail/core/utils.py

```diff
@@ -356,3 +366,145 @@ def multigetattr(item, accessor):
             current = current()
 
     return current
+
+
+def get_dummy_request(path: str = "/", site: 'Site' = None) -> HttpRequest:
+    """
+    Return a simple ``HttpRequest`` instance that can be passed to
+    ``Page.get_url()`` and other methods to benefit from improved performance
+    when no real ``HttpRequest`` instance is available.
+
+    If ``site`` is provided, the ``HttpRequest`` is made to look like it came
+    from that Wagtail ``Site``.
+    """
```

Comment:
Have been doing some code archaeology to figure out why we need to jump through these hoops, and whether we could massively simplify this by constructing a generic `HttpRequest` up-front in `create_redirects` and passing that as `cache_object`.

It comes down to a bad design decision in #4390 - the `request` argument to `get_url_parts` is being assigned a meaning that it was never meant to have [according to the API reference](https://docs.wagtail.io/en/stable/reference/pages/model_reference.html#wagtail.core.models.Page.get_url_parts) - namely, it's assuming that we're building a URL for the purpose of rendering it into a page template, and `request` is the request that initiated that template response.

The one and only time when it comes into play is when there are multiple sites pointing to the same root page, so rather than selecting the deepest-rooted one as usual, it will break the tie by preferring the one that matches `request`'s site.

(I'm discounting the possibility that a page model might override `get_url_parts` to do its own custom `request`-dependent logic, because that clearly goes against the instructions in the `get_url_parts` docs.)

So, unwinding a bit: the only time `_page_urls_for_sites` will ever return multiple results is if the `get_url_parts` result varies on request, and that only happens if this 'multiple sites pointing to the same root' situation is in effect. And in that case, the multiple results returned from `_page_urls_for_sites` will be something of the form

    {
        (first_aliased_site, '/foo', '/bar'),
        (second_aliased_site, '/foo', '/bar'),
    }

where first_aliased_site and second_aliased_site are the multiple sites pointing to the same root, and the other fields are identical.

In short: we only need these specially-crafted request objects to deal with one annoying edge case, and in that edge case, it should be fairly easy to derive the intended results from a single `get_url_parts` call with a generic request. However, by now I think the reasoning to get there is more complicated than the code we've got here, so let's stick with the current version for now - I will personally take on the technical debt and deal with that in a future PR :-)